Sources/NIOCore/ChannelHandler.swift

```diff
@@ -353,6 +353,36 @@ public enum NIOProtocolNegotiationResult<NegotiationResult: Sendable> {
     case deferredResult(EventLoopFuture<NIOProtocolNegotiationResult<NegotiationResult>>)
 }
 
+@_spi(AsyncChannel)
+extension NIOProtocolNegotiationResult {
+    fileprivate func resolve(on eventLoop: EventLoop) -> EventLoopFuture<NegotiationResult> {
+        Self.resolve(on: eventLoop, result: self)
+    }
+
+    fileprivate static func resolve(on eventLoop: EventLoop, result: Self) -> EventLoopFuture<NegotiationResult> {
+        switch result {
+        case .finished(let negotiationResult):
+            return eventLoop.makeSucceededFuture(negotiationResult)
+
+        case .deferredResult(let future):
+            return future.flatMap { result in
+                return resolve(on: eventLoop, result: result)
+            }
+        }
+    }
+}
+
+extension EventLoopFuture {
+    @available(macOS 10.15, *)
+    @_spi(AsyncChannel)
+    /// Resolves the protocol negotiation result.
+    ///
+    /// This method waits for the final protocol negotiation result and the returns the value.
+    public func resolve<NegotiationResult: Sendable>() async throws -> NegotiationResult where Value == NIOProtocolNegotiationResult<NegotiationResult> {
+        try await self.flatMap { $0 .resolve(on: self.eventLoop) }.get()
+    }
```

Comment:
I am not 100% sold on this yet. It is pretty hard to discover. We could instead introduce a new like `NIOUnresolvedProtocolNegotiationResult` which would hold this method and make the `NIOProtocolNegotiationHandler` require to return that instead of the current `EventLoopFuture<NIOProtocolNegotiationResult<NegotiationResult>>`